File: backend/src/services/datastatus.ts
import dayjs from 'dayjs';
import type { Instrument, InstrumentDataStatus } from '@decisionguru/shared';
import { db } from '../db/index.js';

const quoteStmt = db.prepare('SELECT price, fetchedAt FROM quote_cache WHERE symbol = ?');
const coverageStmt = db.prepare(
  'SELECT COUNT(*) AS c, MAX(date) AS mx FROM price_cache WHERE symbol = ?',
);

/**
 * Market-data health for one instrument, used to render a per-instrument badge so the UI
 * never silently shows a 0 as if it were a real value.
 */
export function instrumentDataStatus(
  inst: Instrument,
  openQuantity?: number,
): InstrumentDataStatus {
  if (inst.unresolved || (inst.isin && inst.symbol === inst.isin)) {
    return {
      state: 'unresolved',
      resolutionSource: inst.resolutionSource ?? 'unresolved',
      message: 'No ticker resolved — value & charts unavailable. Edit the symbol or retry resolve.',
    };
  }
  // A fully-closed position has a definitive value of 0 — that is not "missing data".
  if (openQuantity != null && openQuantity <= 0) {
    return { state: 'ok', resolutionSource: inst.resolutionSource ?? null, message: 'Position closed.' };
  }
  const quote = quoteStmt.get(inst.symbol) as { price: number; fetchedAt: number } | undefined;
  const cov = coverageStmt.get(inst.symbol) as { c: number; mx: string | null };
  const hasQuote = !!quote && quote.price > 0;
  const coverageDays = cov.c ?? 0;

  if (!hasQuote && !coverageDays) {
    return {
      state: 'no-data',
      resolutionSource: inst.resolutionSource ?? null,
      priceCoverageDays: 0,
      message: 'No market data yet (provider may be rate-limiting) — try again shortly.',
    };
  }

  const quoteAsOf = quote ? new Date(quote.fetchedAt).toISOString() : null;
  const stale =
    (!quote || dayjs().diff(dayjs(quote.fetchedAt), 'hour') >= 24) &&
    (!cov.mx || dayjs().diff(dayjs(cov.mx), 'day') > 4);

  return {
    state: stale ? 'stale' : 'ok',
    resolutionSource: inst.resolutionSource ?? null,
    quoteAsOf,
    priceCoverageDays: coverageDays,
    message: stale ? 'Prices may be a few days old (cached).' : null,
  };
}

File: backend/app/services/datastatus.py
"""Per-instrument market-data health badge — port of datastatus.ts."""
from __future__ import annotations

import pandas as pd

from ..core import db
from ..core.timefmt import iso_from_ms


def instrument_data_status(inst: dict, open_quantity: float | None = None) -> dict:
    if inst.get("unresolved") or (inst.get("isin") and inst.get("symbol") == inst.get("isin")):
        return {
            "state": "unresolved",
            "resolutionSource": inst.get("resolutionSource") or "unresolved",
            "message": "No ticker resolved — value & charts unavailable. "
                       "Edit the symbol or retry resolve.",
        }
    # A fully-closed position has a definitive value of 0 — not "missing data".
    if open_quantity is not None and open_quantity <= 0:
        return {"state": "ok", "resolutionSource": inst.get("resolutionSource"),
                "message": "Position closed."}

    symbol = inst["symbol"]
    quote = db.q("SELECT price, fetchedAt FROM quote_cache WHERE symbol = ?").get((symbol,))
    cov = db.q("SELECT COUNT(*) AS c, MAX(date) AS mx FROM price_cache WHERE symbol = ?").get((symbol,))
    has_quote = bool(quote) and quote["price"] > 0
    coverage_days = (cov["c"] if cov else 0) or 0

    if not has_quote and not coverage_days:
        return {
            "state": "no-data",
            "resolutionSource": inst.get("resolutionSource"),
            "priceCoverageDays": 0,
            "message": "No market data yet (provider may be rate-limiting) — try again shortly.",
        }

    quote_as_of = iso_from_ms(quote["fetchedAt"]) if quote else None
    now = pd.Timestamp.utcnow().tz_localize(None)
    quote_age_h = (
        (now - pd.Timestamp(quote["fetchedAt"], unit="ms")).total_seconds() / 3600
        if quote else None
    )
    cov_age_d = (
        (now.normalize() - pd.Timestamp(cov["mx"]).normalize()).days
        if cov and cov["mx"] else None
    )
    stale = (
        (quote is None or (quote_age_h is not None and quote_age_h >= 24))
        and (cov is None or not cov["mx"] or (cov_age_d is not None and cov_age_d > 4))
    )

    return {
        "state": "stale" if stale else "ok",
        "resolutionSource": inst.get("resolutionSource"),
        "quoteAsOf": quote_as_of,
        "priceCoverageDays": coverage_days,
        "message": "Prices may be a few days old (cached)." if stale else None,
    }
